File: packages/server/src/queue.ts
// Bounded in-process event queue with exponential-backoff retry and jitter.

export interface QueuedEvent {
  writeKey: string;
  events: unknown[];
  attempt: number;
  nextRetryAt: number;
}

interface QueueOptions {
  maxSize?: number;
  maxRetries?: number;
  baseRetryMs?: number;
}

export class EventQueue {
  private readonly queue: QueuedEvent[] = [];
  private readonly maxSize: number;
  private readonly maxRetries: number;
  private readonly baseRetryMs: number;

  constructor(opts: QueueOptions = {}) {
    this.maxSize = opts.maxSize ?? 1000;
    this.maxRetries = opts.maxRetries ?? 5;
    this.baseRetryMs = opts.baseRetryMs ?? 2000;
  }

  enqueue(item: Omit<QueuedEvent, 'attempt' | 'nextRetryAt'>): boolean {
    if (this.queue.length >= this.maxSize) return false;
    this.queue.push({ ...item, attempt: 0, nextRetryAt: 0 });
    return true;
  }

  dequeueReady(): QueuedEvent | undefined {
    const now = Date.now();
    const idx = this.queue.findIndex((item) => item.nextRetryAt <= now);
    if (idx === -1) return undefined;
    return this.queue.splice(idx, 1)[0];
  }

  requeue(item: QueuedEvent): void {
    if (item.attempt >= this.maxRetries) return;
    const jitter = Math.random() * 1000;
    const delay = this.baseRetryMs * Math.pow(2, item.attempt) + jitter;
    this.queue.push({ ...item, attempt: item.attempt + 1, nextRetryAt: Date.now() + delay });
  }

  get size(): number {
    return this.queue.length;
  }

  drain(): void {
    this.queue.length = 0;
  }
}

File: Backend Architecture/aether-backend/shared/server_sdk/queue.py
"""Bounded in-process event queue with exponential-backoff retry."""

from __future__ import annotations

import time
import random
from dataclasses import dataclass, field


@dataclass
class QueuedBatch:
    events: list[dict]
    attempt: int = 0
    next_retry_at: float = field(default_factory=time.time)


class EventQueue:
    def __init__(self, max_size: int = 1000, max_retries: int = 5, base_retry_s: float = 2.0) -> None:
        self._queue: list[QueuedBatch] = []
        self._max_size = max_size
        self._max_retries = max_retries
        self._base_retry_s = base_retry_s

    def enqueue(self, events: list[dict]) -> bool:
        if len(self._queue) >= self._max_size:
            return False
        self._queue.append(QueuedBatch(events=events))
        return True

    def dequeue_ready(self) -> QueuedBatch | None:
        now = time.time()
        for i, item in enumerate(self._queue):
            if item.next_retry_at <= now:
                return self._queue.pop(i)
        return None

    def requeue(self, item: QueuedBatch) -> None:
        if item.attempt >= self._max_retries:
            return
        jitter = random.random()
        delay = self._base_retry_s * (2 ** item.attempt) + jitter
        self._queue.append(QueuedBatch(
            events=item.events,
            attempt=item.attempt + 1,
            next_retry_at=time.time() + delay,
        ))

    @property
    def size(self) -> int:
        return len(self._queue)

    def drain(self) -> None:
        self._queue.clear()
